### PR TITLE
Scheduling plugin not updating responding attendee status

### LINKED ISSUE #2861: Participants, once accepted, won't show up as accepted in the event.
### Steps to reproduce
1 Create an event with at least an attendee!
2 Let that attendee to accept participating in the event!
3 See an email that was sent that the attendee accepted!
4 See the event and the attendee still has a question mark!

### Expected behaviour
The event shows acceptances.

### Actual behaviour
The event doesn't show acceptances. See the email content:

"Kristel Tamm <kristel.tamm@peetri.edu.ee> has accepted your event invitation.

Title:
	

Accepted: "Covid-19": 6.E (Kristel T)

Location:
	

B148

When:
	

neljapäev, 4. veebruar 2021 09:25 – 10:10

Attendees:
	

	Kristel Tamm <kristel.tamm@peetri.edu.ee>"

Once I open the event, it still shows the question mark for her.

## Server configuration detail

**Operating system:** Linux 5.4.0-64-generic #72-Ubuntu SMP Fri Jan 15 10:27:54 UTC 2021 x86_64

**Webserver:** Apache/2.4.41 (Ubuntu) (apache2handler)

**Database:** mysql 10.3.25

**PHP version:** 

7.4.3
Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, session, standard, sodium, apache2handler, mysqlnd, PDO, xml, apcu, bcmath, calendar, ctype, curl, dom, mbstring, FFI, fileinfo, ftp, gd, gettext, gmp, iconv, imagick, imap, intl, json, exif, mysqli, pdo_mysql, apc, posix, readline, shmop, SimpleXML, sockets, sysvmsg, sysvsem, sysvshm, tokenizer, xmlreader, xmlwriter, xsl, zip, Phar, Zend OPcache

**Nextcloud version:** 20.0.6 - 20.0.6.1

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** unknown

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.6.0
 - activity: 2.13.4
 - admin_audit: 1.10.0
 - audioplayer: 3.0.0
 - bruteforcesettings: 2.0.1
 - calendar: 2.1.3
 - cloud_federation_api: 1.3.0
 - comments: 1.10.0
 - contacts: 3.4.3
 - contactsinteraction: 1.1.0
 - dashboard: 7.0.0
 - dav: 1.16.2
 - deck: 1.2.3
 - federatedfilesharing: 1.10.2
 - federation: 1.10.1
 - files: 1.15.0
 - files_external: 1.11.1
 - files_pdfviewer: 2.0.1
 - files_rightclick: 0.17.0
 - files_sharing: 1.12.2
 - files_texteditor: 2.14.0
 - files_trashbin: 1.10.1
 - files_versions: 1.13.0
 - files_videoplayer: 1.9.0
 - firstrunwizard: 2.9.0
 - forms: 2.1.0
 - gpxmotion: 0.1.0
 - issuetemplate: 0.7.0
 - logreader: 2.5.0
 - lookup_server_connector: 1.8.0
 - mail: 1.8.0
 - maps: 0.1.8
 - metadata: 0.12.0
 - nextcloud_announcements: 1.9.0
 - notifications: 2.8.0
 - oauth2: 1.8.0
 - onlyoffice: 2.1.2
 - password_policy: 1.10.1
 - photos: 1.2.3
 - polls: 1.7.4
 - privacy: 1.4.0
 - provisioning_api: 1.10.0
 - recommendations: 0.8.0
 - registration: 0.6.0
 - scanner: 0.2.0
 - serverinfo: 1.10.0
 - settings: 1.2.0
 - sharebymail: 1.10.0
 - support: 1.3.0
 - survey_client: 1.8.0
 - systemtags: 1.10.0
 - tasks: 0.13.6
 - text: 3.1.0
 - theming: 1.11.0
 - twofactor_backupcodes: 1.9.0
 - updatenotification: 1.10.0
 - user_status: 1.0.1
 - viewer: 1.4.0
 - weather: 1.7.4
 - weather_status: 1.0.0
 - wopi: 3.5.6
 - workflowengine: 2.2.0
Disabled:
 - dicomviewer
 - drawio
 - drop_account
 - emlviewer
 - encryption
 - external
 - files_automatedtagging
 - files_external_gdrive
 - ojsxc
 - richdocuments
 - spreed
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "nextcloud.tennis24.ee"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "20.0.6.1",
    "overwrite.cli.url": "https:\/\/nextcloud.tennis24.ee",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "maintenance": false,
    "theme": "",
    "loglevel": 0,
    "app_install_overwrite": [
        "onlyoffice",
        "dashboard",
        "files_external_gdrive",
        "dicomviewer",
        "calendar"
    ],
    "tempdirectory": "\/projektid\/nextcloud\/data",
    "overwriteprotocol": "https",
    "mysql.utf8mb4": true,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "has_rebuilt_cache": true,
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpmode": "smtp",
    "mail_sendmailmode": "smtp",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauthtype": "LOGIN",
    "mail_smtpauth": 1,
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "587",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpsecure": "tls"
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** 

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

## Client configuration

**Browser:** Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36

**Operating system:** 

## Logs

<details><summary>Web server error log</summary>

```
Insert your web server log here 
```
</details>

<details><summary>Nextcloud log</summary>

```
Insert your Nextcloud log here
```
</details>

<details><summary>Browser log</summary>

Insert your browser log here, this could for example include:

	a) The javascript console log
	b) The network log
	c) ...
</details>

### PR DESCRIPTION
**Pull request by miaulalala**
Fixes https://github.com/nextcloud/calendar/issues/2861

### CODE REVIEW

**Comment by miaulalala** [comment]
Shall we backport this?

**Comment by miaulalala** [comment]
Drone failure is unrelated

File: apps/dav/lib/CalDAV/Schedule/Plugin.php [hunk 1]
```diff
@@ -554,4 +567,155 @@ private function stripOffMailTo(string $email): string {
...
+			if ('REPLY' === $iTipMessage->method) {
+				$this->processICalendarChange(
+					$oldICalendarData,
+					$newObject,
+					[$iTipMessage->recipient],
+					// this used to have the sender in the ignore field
+					// removed that because creating a scheduling update for the
+					// attendee would mean duplicating all this code
+					// this is not RFC6638 conform (See Section 4.2), but it's working
+					[]
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/Schedule/Plugin.php:714]: @kesselb here is the only change compared to Sabre

**Comment by kesselb** [comment]
I'm not able to reproduce the initial issue on master anymore. On master and this branch the attendance status is updated properly and shows the right state :confused:

**Comment by kesselb** [comment]
> I'm not able to reproduce the initial issue on master anymore. On master and this branch the attendance status is updated properly and shows the right state confused

My bad :see_no_evil:  I looked at the wrong point. For master and this pull request the attendance status for the organizer is updated properly. 

The purpose of this pull request is to fix an issue with updating the attendance status for the attendee itself. To reproduce create a event and invite Alice and Bob. Wait for Bob's invitation email and accept the invite. The attendance status for the organizer and Alice is updated (Bob accepted the invite). In Bob's calendar event the attendance status is still pending. 

A possible explanation for this case is the way we handle event invitations. If a user accept or decline an invitation a iTip message is generated. 

https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/apps/dav/lib/Controller/InvitationResponseController.php#L200-L201

The recipient is used to select the right event from the calendar. When Alice accept the invite an iTip message is generated with sender = alice@example.org and recipient = organizer@example.org. The event for recipient organizer@example.org is selected and the attendance status for alice@example.org updated. Then for every other attendee another iTip message is generated to also update their event. 

https://github.com/nextcloud/3rdparty/blob/6876f1fce8d1c70790c165dd7ed0b4214364e397/sabre/dav/lib/CalDAV/Schedule/Plugin.php#L527-L538

image

Sabre assume that we already updated Alice event and there is no need to sent a iTip message to Alice. 

```
$iTipMessage->sender = $row['attendee']; 
$iTipMessage->recipient = $row['attendee']; 
```

I'm not sure if that's a valid iTip message. But when sender = recipient the calendar event for the attendee is updated and a iTip message for the organizer and other attendees generated. 

Just adding my findings here to document it for later. I guess it's fine to merge this pull request and take out the ignore check.

**Review by kesselb — APPROVED** [review]

**Comment by miaulalala** [comment]
Drone failure unrelated.

**Comment by miaulalala** [comment]
This has been discussed an the actual RFC compliant way will be to update the recipient in the iTip Message. Thanks @kesselb for figuring this out!

**Review by st3iny — APPROVED** [review]
I tested it and the status was updated correctly for both the organizer and attendee. I didn't look at the code though.

**Pull request closed without merging**